src/core/lib/gpr/arena.cc

```diff
@@ -118,35 +122,49 @@ void* gpr_arena_alloc(gpr_arena* arena, size_t size) {
-      size_t next_z_size =
-          static_cast<size_t>(gpr_atm_no_barrier_load(&arena->size_so_far));
-      next_z = static_cast<zone*>(zalloc_aligned(
-          GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone)) + next_z_size));
-      next_z->size_begin = z->size_end;
-      next_z->size_end = z->size_end + next_z_size;
-      if (!gpr_atm_rel_cas(&z->next_atm, static_cast<gpr_atm>(NULL),
-                           (gpr_atm)next_z)) {
-        gpr_free_aligned(next_z);
-        next_z = (zone*)gpr_atm_acq_load(&z->next_atm);
+
+  // Check to see if the allocation isn't able to start in the initial zone.
+  // This statement is true only in the uncommon case because of our arena
+  // sizing historesis (that is, most calls should have a large enough initial
+  // zone and will not need to grow the arena).
+  if (start >= z->size_end) {
+    // Find the correct zone to start the allocation in
+    gpr_mu_lock(&arena->arena_growth_mutex);
+    while (start >= z->size_end) {
+      if (z->next == nullptr) {
+        size_t next_z_size = total_size_of_arena_allocations;
+        z->next = static_cast<zone*>(zalloc_aligned(
+            GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone)) + next_z_size));
+        z->next->size_begin = z->size_end;
+        z->next->size_end = z->size_end + next_z_size;
       }
+      z = z->next;
     }
-    z = next_z;
+    gpr_mu_unlock(&arena->arena_growth_mutex);
   }
-  if (start + size > z->size_end) {
+  // Now that we have a zone we could start in, can the allocation fit in this
+  // zone? If not, call arena_alloc again. Note that this means some space in
+  // the arena is wasted because we have already incremented the size_so_far of
+  // the arena and we'll increment it again for the same attempted allocation.
+  if (total_size_of_arena_allocations > z->size_end) {
     return gpr_arena_alloc(arena, size);
   }
   GPR_ASSERT(start >= z->size_begin);
-  GPR_ASSERT(start + size <= z->size_end);
-  char* ptr = (z == &arena->initial_zone)
-                  ? reinterpret_cast<char*>(arena) +
-                        GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(gpr_arena))
-                  : reinterpret_cast<char*>(z) +
-                        GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone));
-  return ptr + start - z->size_begin;
+  GPR_ASSERT(total_size_of_arena_allocations <= z->size_end);
+  // Skip the first part of the zone, which just contains tracking information.
+  // For the initial zone, this is the gpr_arena struct and for any other zone,
+  // it's the zone struct.
+  char* ptr_to_alloc_location =
+      (z == &arena->initial_zone)
+          ? reinterpret_cast<char*>(arena) +
+                GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(gpr_arena))
+          : reinterpret_cast<char*>(z) +
+                GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone));
+  // start - size_begin is how many bytes have been allocated into the current
+  // zone
+  return ptr_to_alloc_location + start - z->size_begin;
```

Comment:
Does this mean that whever there is more than one zone, we waste the initial space in the each subsequent zone equal to the size of all preceding zones?  Or am I reading this wrong?

Reply:
> `z->size_begin` is the sum of all the allocations on the arena so far up until zone z.

I don't think you mean the sum of all allocations in the arena so far up until zone z; I think you mean the size of all zones preceding zone z.  Is that right?

The rest of this explanation makes sense, thanks!

Reply:
Okay, yeah, I think we're saying the same thing.  I was assuming that the term "allocations" did not include the waste, which was the source of my confusion.  And I was using the phrase "size of all zones" to not include the tracking information for each zone.

Terminology is hard. :)